### PR TITLE
allow expand op and adjust vectorization check in memory alias

### PR DESCRIPTION
**Pull request by liqiangxl**
Fix shared memory reuse in softmax by:
(1) allow `expand` op and treat it similar to broadcast, that is `require matched iterdomains for reuse`
(2) Allow non-vectorized tv to reuse vectorized tv.

see https://github.com/NVIDIA/Fuser/issues/633 and https://github.com/NVIDIA/Fuser/issues/1562

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build --diff --diff-bench

**Comment by liqiangxl** [comment]
The change of `Allow non-vectorized tv to reuse vectorized tv` added explicity register alias in some of the matmul [cases](https://dl.gitlab-master-pages.nvidia.com/pytorch/fuser-gh-mirror//nvfuser_github_ci/codegen_diff_p11711578_j78194308_1703573108667311342_codediff_a3d1e61_4acaf81_custom_command_20231226_055718.html), and `Softmax_Inner_fp32`, see codegen diffs.

**Comment by naoyam** [comment]
Is the alignment check necessary for smem? I thought we always align smem allocations at 16B boundaries.

**Comment by liqiangxl** [comment]
> Is the alignment check necessary for smem? I thought we always align smem allocations at 16B boundaries.

You are right. We can skip this check for smem since the largest vectorization is 16 Bytes.

File: csrc/device_lower/pass/alias_memory.cpp [hunk 1]
```diff
@@ -1202,21 +1202,19 @@ class ReusableAllocationFinder : private kir::IrVisitor {
...
+          // Vectorized allocations require correct alignment so if [this_tv]
+          // is vectorized, the [reuse_tv] must be vectorized with the same
+          // factor.
+          // No need to check shared memory since it is always aligned to 16
+          // Bytes which is also the maximum vectorization width.
+          if (this_tv->getMemoryType() == MemoryType::Local) {
+            const auto& va = GpuLower::current()->vectorizedAccesses();
+            if ((va.find(this_tv) != va.end())) {
+              if ((va.find(reuse_tv) == va.end()) ||
+                  va.at(this_tv) != va.at(reuse_tv)) {
```
Comment (naoyam) [inline csrc/device_lower/pass/alias_memory.cpp:1215]: Let's make the code simpler and more obvious:

```suggestion
bool this_tv_vectorized = va.find(this_tv) != va.end();
if (this_tv_vectorized) {
  bool reuse_tv_vectorized = va.find(reuse_tv) != va.end();
  if (!reuse_tv_vectorized) {
   return false;
  }
  int this_tv_alignment = va.at(this_tv);
  int reuse_tv_alignment = va.at(reuse_tv);
  if (this_tv_alignment != reuse_tv_alignment) {
    return false;
  }
}
```

Also can we relax the alignment check as below??
```
  if (this_tv_alignment > reuse_tv_alignment) {
    return false;
  }
```

**Comment by naoyam** [comment]
Please add unit tests

**Comment by naoyam** [comment]
Also, it would be good if we could have better protection or error detection from the wrong estimation of smem usage. I understand that we need to estimate the smem usage before actually scheduling a fusion, but can we actually validate the estimate after the scheduling (but before lowering)?

File: csrc/device_lower/pass/alias_memory.cpp (hunk 1 above)
Comment (liqiangxl) [reply csrc/device_lower/pass/alias_memory.cpp]: changed.

**Comment by liqiangxl** [comment]
> Please add unit tests

added 2 tests, one for smem reuse, the other for register reuse with varied vectorization factors.

**Comment by liqiangxl** [comment]
> estimation of smem usage.

smem/register reuse before scheduling is only based on lifescope analysis. However the actual reuse in `reuseMemoryAllocations` depends on how the fusion is actually scheduled. It seems difficult if possible to validate the estimated memory usage before lowering. Maybe create another issue?

**Comment by liqiangxl** [comment]
> Also, it would be good if we could have better protection or error detection from the wrong estimation of smem usage. I understand that we need to estimate the smem usage before actually scheduling a fusion, but can we actually validate the estimate after the scheduling (but before lowering)?

smem/register reuse before scheduling is only based on lifescope analysis. However the actual reuse in reuseMemoryAllocations depends on how the fusion is actually scheduled. It seems difficult if possible to validate the estimated memory usage before lowering. Maybe create another issue?

File: test/test_smem_reuse.cpp [hunk 2]
```diff
@@ -511,4 +511,122 @@ TEST_F(SmemReuseTest, SkippedSyncInterval) {
...
+    testValidate(fusion.get(), cg_outputs, {t0}, __LINE__, __FILE__);
+  };
+
+  // test different vectorization factors, max is 4 since float is used.
+  for (int vect_factor_1 : {1, 2, 4}) {
+    for (int vect_factor_2 : {1, 2, 4}) {
+      testRegisterReuse(vect_factor_1, vect_factor_2);
+    }
+  }
+}
```
Comment (jacobhinkle) [inline test/test_smem_reuse.cpp:631]: Can we add a test with an `ExpandOp` that interferes with re-use here?

**Review by jacobhinkle — COMMENTED** [review]
The changes make sense to me. `ExpandOp` is being grouped with `BroadcastOp` here, but I think it's really a no-op wrt alias analysis since it cannot introduce broadcasts on its own. I could be wrong but my understanding is that `ExpandOp` really is just constraining the extents of other `TensorView`s that might resolve a broadcast, but I don't think it actually changes the indexing of its argument, so it should basically be ignored in this alias analysis.

**Comment by naoyam** [comment]
> > Also, it would be good if we could have better protection or error detection from the wrong estimation of smem usage. I understand that we need to estimate the smem usage before actually scheduling a fusion, but can we actually validate the estimate after the scheduling (but before lowering)?
> 
> smem/register reuse before scheduling is only based on lifescope analysis. However the actual reuse in reuseMemoryAllocations depends on how the fusion is actually scheduled. It seems difficult if possible to validate the estimated memory usage before lowering. Maybe create another issue?

Hmm, we already know which tensors are cached on the shared memory, so isn't it just checking their leaf domains and the computeAt positions?

**Comment by liqiangxl** [comment]
> > > Also, it would be good if we could have better protection or error detection from the wrong estimation of smem usage. I understand that we need to estimate the smem usage before actually scheduling a fusion, but can we actually validate the estimate after the scheduling (but before lowering)?
> > 
> > 
> > smem/register reuse before scheduling is only based on lifescope analysis. However the actual reuse in reuseMemoryAllocations depends on how the fusion is actually scheduled. It seems difficult if possible to validate the estimated memory usage before lowering. Maybe create another issue?
> 
> Hmm, we already know which tensors are cached on the shared memory, so isn't it just checking their leaf domains and the computeAt positions?

If just check leaf domains and the computeAt positions, we can do that after schedule. I am not sure if this is equivalent to the checks in reuseMemoryAllocations. 
I added `validateSmemPersistentBufferSize` at the end of `schedulePersistentKernel`, It works as follows:
```
// If there are multiple persistent buffers
// and not all of them are active at the same program point,
// It indicates shared memory reuse and needs to validate the reuse.
// Find the program point where the number of active buffers is maximum,
// then check the reuse of inactive buffers by active buffers.
// The reuse pair must have the same leaf domains (except for vectorization)
// and compute at position.
```

File: test/test_smem_reuse.cpp (hunk 2 above)
Comment (liqiangxl) [reply test/test_smem_reuse.cpp:631]: added

File: csrc/device_lower/pass/alias_memory.cpp [hunk 3]
```diff
@@ -1341,7 +1345,7 @@ class ReusableAllocationFinder : private kir::IrVisitor {
         }
         if (!ir_utils::isPointwiseTvOp(tv_def) &&
             !ir_utils::isReductionTvOp(tv_def)) {
```
[the remark this answers is not in the document]
Comment (liqiangxl) [reply csrc/device_lower/pass/alias_memory.cpp]: changed.

**Comment by naoyam** [comment]
> > > > Also, it would be good if we could have better protection or error detection from the wrong estimation of smem usage. I understand that we need to estimate the smem usage before actually scheduling a fusion, but can we actually validate the estimate after the scheduling (but before lowering)?
> > > 
> > > 
> > > smem/register reuse before scheduling is only based on lifescope analysis. However the actual reuse in reuseMemoryAllocations depends on how the fusion is actually scheduled. It seems difficult if possible to validate the estimated memory usage before lowering. Maybe create another issue?
> > 
> > 
> > Hmm, we already know which tensors are cached on the shared memory, so isn't it just checking their leaf domains and the computeAt positions?
> 
> If just check leaf domains and the computeAt positions, we can do that after schedule. I am not sure if this is equivalent to the checks in reuseMemoryAllocations. I added `validateSmemPersistentBufferSize` at the end of `schedulePersistentKernel`, It works as follows:
> 
> ```
> // If there are multiple persistent buffers
> // and not all of them are active at the same program point,
> // It indicates shared memory reuse and needs to validate the reuse.
> // Find the program point where the number of active buffers is maximum,
> // then check the reuse of inactive buffers by active buffers.
> // The reuse pair must have the same leaf domains (except for vectorization)
> // and compute at position.
> ```

Oh, I see. Yeah, you're right, the actual used size depends on the reuse logic, which is still unknown at the point of the scheduling.

Thanks for adding the validation routine, but it doesn't seem add much value as the check is mostly just about what the schedule does. Let's not add the check for now.

**Comment by naoyam** [comment]
The PR almost looks good to me. See my previous comment (https://github.com/NVIDIA/Fuser/pull/1561#issuecomment-1874973026). I'll let @jacobhinkle give a final stamp.

**Comment by liqiangxl** [comment]
> validate reuse after schedule before lowering



> > > > > Also, it would be good if we could have better protection or error detection from the wrong estimation of smem usage. I understand that we need to estimate the smem usage before actually scheduling a fusion, but can we actually validate the estimate after the scheduling (but before lowering)?
> > > > 
> > > > 
> > > > smem/register reuse before scheduling is only based on lifescope analysis. However the actual reuse in reuseMemoryAllocations depends on how the fusion is actually scheduled. It seems difficult if possible to validate the estimated memory usage before lowering. Maybe create another issue?
> > > 
> > > 
> > > Hmm, we already know which tensors are cached on the shared memory, so isn't it just checking their leaf domains and the computeAt positions?
> > 
> > 
> > If just check leaf domains and the computeAt positions, we can do that after schedule. I am not sure if this is equivalent to the checks in reuseMemoryAllocations. I added `validateSmemPersistentBufferSize` at the end of `schedulePersistentKernel`, It works as follows:
> > ```
> > // If there are multiple persistent buffers
> > // and not all of them are active at the same program point,
> > // It indicates shared memory reuse and needs to validate the reuse.
> > // Find the program point where the number of active buffers is maximum,
> > // then check the reuse of inactive buffers by active buffers.
> > // The reuse pair must have the same leaf domains (except for vectorization)
> > // and compute at position.
> > ```
> 
> Oh, I see. Yeah, you're right, the actual used size depends on the reuse logic, which is still unknown at the point of the scheduling.
> 
> Thanks for adding the validation routine, but it doesn't seem add much value as the check is mostly just about what the schedule does. Let's not add the check for now.

The commit adding validation routine is being removed.

**Comment by liqiangxl** [comment]
!build --diff --diff-bench

File: csrc/device_lower/pass/alias_memory.cpp [hunk 4]
```diff
@@ -1381,7 +1385,7 @@ class ReusableAllocationFinder : private kir::IrVisitor {
     }
   }
 
-  // Utility to capture broadcast ops
+  // Utility to capture broadcast and expand ops
```
Comment (jacobhinkle) [inline csrc/device_lower/pass/alias_memory.cpp:1388]:
```suggestion
  // Utility to capture broadcast ops
```

**Review by jacobhinkle — APPROVED** [review]
LGTM!
